### PR TITLE
[RLlib] Issue 18812: Torch multi-GPU stats not protected against race conditions.

### LINKED ISSUE #18812: [Bug] [RLLIB] Race condition in stats_fn when using multi-gpu
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

There is a race condition for the stats_fn when using multi-gpu.
I only tested with torch and ppo but it will also occur with tf.
This will cause an issue for any algorithm in which the loss function stores values as members in the policy for later retrieval.

The crux of the issue is that some algorithms store values in the policy object for retrieval later. PPO for example stores the "mean_kl" value that is used to update the kl_coeff here. 

https://github.com/ray-project/ray/blob/39bcbe03bcaadec3233216dc890759e4a48a38cb/rllib/agents/ppo/ppo_torch_policy.py#L108-L116

The multi-gpu learner is a multi-threaded learner that is computing the loss in batches for each gpu device in parallel. Since they are all accessing the same policy object you have a race condition on which stats get stored.

Here is a sample output from a run I did training cartpole with 2 gpus. You can see that the values stared by the two sample batches in the loss function and the values for the two towers in the policy "learn_on_loaded_batch" function do not match.

```
(pid=3484436) In ppo agent: 
(pid=3484436) device:  cuda:1
(pid=3484436) total_loss:  tensor(1.3728, device='cuda:1', grad_fn=<DivBackward0>)
(pid=3484436) mean_policy_loss:  tensor(0.5660, device='cuda:1', grad_fn=<DivBackward0>)
(pid=3484436) mean_vf_loss:  tensor(80.3185, device='cuda:1', grad_fn=<DivBackward0>)
(pid=3484436) mean_entropy:  tensor(0.6534, device='cuda:1', grad_fn=<DivBackward0>)
(pid=3484436) mean_kl:  tensor(0.0179, device='cuda:1', grad_fn=<DivBackward0>)

(pid=3484436) In ppo agent: 
(pid=3484436) device:  cuda:0
(pid=3484436) total_loss:  tensor(4.2556, device='cuda:0', grad_fn=<DivBackward0>)
(pid=3484436) mean_policy_loss:  tensor(-0.5995, device='cuda:0', grad_fn=<DivBackward0>)
(pid=3484436) mean_vf_loss:  tensor(485.2775, device='cuda:0', grad_fn=<DivBackward0>)
(pid=3484436) mean_entropy:  tensor(0.6738, device='cuda:0', grad_fn=<DivBackward0>)
(pid=3484436) mean_kl:  tensor(0.0118, device='cuda:0', grad_fn=<DivBackward0>)

(pid=3484436) In torch policy:
{'tower_0': {'learner_stats': {'cur_kl_coeff': 0.2, 'cur_lr': 0.0003, 'total_loss': 1.3728129863739014, 'policy_loss': 0.5660414099693298, 'vf_loss': 80.31847381591797, 'vf_explained_var': array([-0.00416613], dtype=float32), 'kl': 0.017934031784534454, 'entropy': 0.6533743739128113, 'entropy_coeff': 0.0}},
'tower_1': {'learner_stats': {'cur_kl_coeff': 0.2, 'cur_lr': 0.0003, 'total_loss': 1.3728129863739014, 'policy_loss': 0.5660414099693298, 'vf_loss': 80.31847381591797, 'vf_explained_var': array([-0.00416613], dtype=float32), 'kl': 0.017934031784534454, 'entropy': 0.6533743739128113, 'entropy_coeff': 0.0}}, 'learner_stats': {}}
```

### Reproduction script

I made to changes to test this. 

On ray/rllib/agents/ppo/ppo_torch_policy.py:117 I added
```python
    if total_loss.device.index is not None:
        import time
        time.sleep(total_loss.device.index*5)
        print("In ppo agent: ")
        print("device: ", total_loss.device)
        print("total_loss: ", total_loss)
        print("mean_policy_loss: ", mean_policy_loss)
        print("mean_vf_loss: ", mean_vf_loss)
        print("mean_entropy: ", mean_entropy)
        print("mean_kl: ", mean_kl)
```
On ray/rllib/policy/torch_policy.py:651I added
```python
print("In torch policy: ", batch_fetches)
```

This is the yaml I used:
```yaml
cartpole-ppo-two-gpus:
    env: CartPole-v0
    run: PPO
    stop:
        episode_reward_mean: 150
        training_iteration: 400
    config:
        # Works for both torch and tf.
        framework: torch
        gamma: 0.99
        lr: 0.0003
        num_workers: 1
        observation_filter: MeanStdFilter
        num_sgd_iter: 6
        vf_loss_coeff: 0.01

        # Double batch size (2 GPUs).
        train_batch_size: 8000

        model:
            # Test w/ LSTMs. Problem also exists without lstms
            use_lstm: true
            fcnet_hiddens: [32]
            fcnet_activation: linear
            vf_share_layers: true

        # Use 2 GPUs.
        num_gpus: 2
```                      

### Anything else

This issue is separate but revealed the problem descibed here: https://github.com/ray-project/ray/issues/18116

CC: @sven1977 @gjoliver 

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by sven1977**
Issue 18812: Torch multi-GPU stats not protected against race conditions.

This PR:
- moves loss stats (which are computed per-tower) to the individual towers (copies of the model).
- in the stats_fn (run after all towers have computed their losses), these stats can now be averaged (or min/max'd) from the individual towers.
- unifies all stats with the already existing handling of td-errors (which need to remain per-batch item)

## Why are these changes needed?

#18812 

## Related issue number

Closes #18812 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sven1977** [comment]
@mvindiola1 ^^

File: rllib/agents/dqn/r2d2_torch_policy.py [hunk 1]
```diff
@@ -279,7 +291,7 @@ def extra_action_out_fn(policy: Policy, input_dict, state_batches, model,
     postprocess_fn=postprocess_nstep_and_prio,
     optimizer_fn=adam_optimizer,
     extra_grad_process_fn=grad_process_and_td_error_fn,
-    extra_learn_fetches_fn=lambda policy: {"td_error": policy._td_error},
+    extra_learn_fetches_fn=concat_multi_gpu_td_errors,
```
Comment (michaelzhiluo) [inline rllib/agents/dqn/r2d2_torch_policy.py:294]: Might be better just to hardcode the lambda function directly as a function in the r2d2 policy class

File: rllib/agents/dqn/simple_q_torch_policy.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from ray.rllib.policy.torch_policy import TorchPolicy
 from ray.rllib.utils.annotations import override
 from ray.rllib.utils.framework import try_import_torch
-from ray.rllib.utils.torch_ops import huber_loss
+from ray.rllib.utils.torch_ops import concat_multi_gpu_td_errors, huber_loss
```
Comment (michaelzhiluo) [inline rllib/agents/dqn/simple_q_torch_policy.py:19]: Think its better not to abstract td_errors away

**Review by michaelzhiluo — APPROVED** [review]
